Visual improvements to the block navigator

adamziel opened the pull request:
## Description

This PR is an initial take on the design changes proposed in  https://github.com/WordPress/gutenberg/issues/22497

<img width="1095" alt="Zrzut ekranu 2020-06-1 o 16 08 13" src="https://user-images.githubusercontent.com/205419/83417242-1bc59b80-a422-11ea-808b-645e2b933115.png">

## What's missing

- [x] Inserter (@noisysocks works on it in #22170)
- [ ] Collapsing (requires more discussion - e.g. how collapse action would be different from select and edit?
- [ ] Custom icons for different types of menu items

adamziel commented on packages/edit-navigation/src/components/menu-editor/style.scss:
```diff
@@ -49,12 +49,40 @@
+	.components-card__header {
+		font-weight: bold;
+		border-bottom: 0;
+		padding-top: $grid-unit-30 !important;
+		padding-bottom: $grid-unit-30 !important;
```
let's get rid of these `!important` before merging

adamziel commented on packages/edit-navigation/src/components/menu-editor/style.scss:
```diff
@@ -49,12 +49,40 @@
 	}
 }
 
+.block-editor-block-navigation-leaf {
+	transition: background 0.1s ease-out;
+	@include reduce-motion("transition");
+
+	> :first-child {
+		padding-left: $grid-unit-20;
+	}
+	> :last-child {
+		padding-right: $grid-unit-20;
+	}
+
+	&:hover {
+		background: $light-gray-300;
+	}
+
+	.block-editor-block-navigation-block-contents {
+		padding-top: $grid-unit-15;
+		padding-bottom: $grid-unit-15;
+	}
+}
+
```
These changes should live in `components/block-navigation/style.scss` if they should also apply to the post editor block navigator.

adamziel commented on packages/block-editor/src/components/block-navigation/style.scss:
```diff
@@ -74,7 +81,7 @@ $tree-item-height: 36px;
 	.block-editor-block-mover-button {
 		width: $button-size;
-		height: $button-size;
+		height: 16px;
```
Movers PR should supersede these customizations

jasmussen replied:
Do you mean this one? https://github.com/WordPress/gutenberg/pull/22673

jasmussen commented:
Before:

![before 2](https://user-images.githubusercontent.com/1204802/83609247-4af11f80-a57e-11ea-98c7-6ed7ff83964b.gif)

After:

![after](https://user-images.githubusercontent.com/1204802/83609253-4dec1000-a57e-11ea-9eef-2d5d8ee30c44.gif)

In short, this is a massive improvement. Well done.

A few small things:

- Can we distinguish between the type of menu item and show the corresponding icon? Page keeps the page icon present now. The page set as "home" gets a house icon. The "Posts" page gets a blog icon. External links get a link icon, and categories get a category icon. I don't think we have a "posts" icon yet, but I'd be more than happy to make one for you if we can.
- Alternatively, we should go back to a generic "menu item" icon. 

One thing we need to fix is the row-height. It's 36px which doesn't afford big enough tap targets:

![Uploading Screenshot 2020-06-03 at 09.47.42.png…]()

We should make it 48px tall, so that it can afford up and down mover buttons that are 24x24 both:

<img width="494" alt="Screenshot 2020-06-03 at 09 43 09" src="https://user-images.githubusercontent.com/1204802/83610150-5db82400-a57f-11ea-945d-469938fa759d.png">

adamziel commented:
@jasmussen wait, the "after" gif you posted looks different that what I see on my end:

![2020-06-03 10-37-12 2020-06-03 10_37_26](https://user-images.githubusercontent.com/205419/83615023-4b8db400-a586-11ea-8ab3-5eb7959218be.gif)

> A few small things

Sure! Looking into these now

jasmussen commented:
Ah no, it's the same, I just tested this in the block navigator that is invoked in the Navigation block, whereas you're showing it in context of the edit-menus screen! Which is good, becasue it should be the same in both.

adamziel committed:
Adjust movers size

adamziel committed:
Restore the generic icon

adamziel commented:
@jasmussen here's how it looks like in the "global" block navigator in the post editor:

<img width="283" alt="Zrzut ekranu 2020-06-3 o 12 46 36" src="https://user-images.githubusercontent.com/205419/83628192-649f6080-a598-11ea-8535-7d67fcb873d8.png">

It seems a bit off, maybe the background shouldn't react to hover there? Or just the button should light up?

adamziel commented:
Hovering just the buttons seems off as well, I think we're better off not highlighting at all in the popover:

<img width="296" alt="Zrzut ekranu 2020-06-3 o 13 04 58" src="https://user-images.githubusercontent.com/205419/83629703-e5f7f280-a59a-11ea-97da-a7253174b205.png">

adamziel committed:
Consistently apply whitespace and highlighting throughout all the navigators

adamziel committed:
Skip the top-level navigation block for navigation block specific block navigator (this naming is so confusing!)

adamziel commented:
@jasmussen 

> Can we distinguish between the type of menu item and show the corresponding icon? Page keeps the page icon present now. The page set as "home" gets a house icon. The "Posts" page gets a blog icon. External links get a link icon, and categories get a category icon. I don't think we have a "posts" icon yet, but I'd be more than happy to make one for you if we can.

That's possible, but I'd rather move it to another PR as it would involve grabbing the ID of the current home and posts pages from the API.

> One thing we need to fix is the row-height. It's 36px which doesn't afford big enough tap targets

It is 48px now, although the movers are visually further apart from each other:

<img width="282" alt="Zrzut ekranu 2020-06-3 o 13 13 56" src="https://user-images.githubusercontent.com/205419/83630435-1a1fe300-a59c-11ea-9eab-bc70281c3f76.png">

<img width="283" alt="Zrzut ekranu 2020-06-3 o 13 10 21" src="https://user-images.githubusercontent.com/205419/83630130-99f97d80-a59b-11ea-8e43-816249c87503.png">

This could be addressed by moving the arrows closer to the button's edge, but I'm not too sure about this approach:

<img width="283" alt="Zrzut ekranu 2020-06-3 o 13 12 59" src="https://user-images.githubusercontent.com/205419/83630359-fe1c4180-a59b-11ea-8223-7197c8d00f04.png">

<img width="288" alt="Zrzut ekranu 2020-06-3 o 13 13 10" src="https://user-images.githubusercontent.com/205419/83630363-ff4d6e80-a59b-11ea-96f9-9af7fadcf5b6.png">

Maybe we should just keep them more spread apart. I'm curious about your opinion.

adamziel commented:
This one is ready for a re-review

adamziel committed:
Use a variable for mover button size

jasmussen commented:
> It seems a bit off, maybe the background shouldn't react to hover there? Or only the button should light up?

Yep, good call, let's unbundle the hover change and I/we can explore that separately.

> That's possible, but I'd rather move it to another PR as it would involve grabbing the ID of the current home and posts pages from the API.

Totally fine, and totally non urgent. Might not need to even ticket it, if it feels necessary the instinct will resurface organically.

> This could be addressed by moving the arrows closer to the button's edge, but I'm not too sure about this approach:

I think it's important we move them closer to each other, and even adjust the focus rectangle to match. The reason it's important is so that they feel like the same component that's used for the redesigned movers that will land any moment now: https://github.com/WordPress/gutenberg/pull/22673

If you check that PR out, you'll note how the mover buttons are much bigger than the focus rectangle, so the tap target stays large. I have some things on my plate this week, but otherwise I'm happy to help adjust the focus rectangles to match.

jasmussen commented:
Nice improvements overall, fast work, thank you.

> This one is ready for a re-review

![nav](https://user-images.githubusercontent.com/1204802/83631131-4ee06a00-a59d-11ea-8e79-6dd95aef7dd2.gif)

You were right, the gray hover needs to be looked at separately, let's remove that for now. 

In comparing the old version, this one:

<img src="https://user-images.githubusercontent.com/1204802/83609253-4dec1000-a57e-11ea-9eef-2d5d8ee30c44.gif">

with the mockup:

<img width="350" alt="More Menu, with lines" src="https://user-images.githubusercontent.com/1204802/83631443-d332ed00-a59d-11ea-825d-6513880a1f05.png">

Something does feel slightly lost with the removal of the tree view for the top level. Not that we should add it back, as soon as you have a submenu item, things fall into place:

<img width="408" alt="Screenshot 2020-06-03 at 13 27 36" src="https://user-images.githubusercontent.com/1204802/83631617-12f9d480-a59e-11ea-81dd-9df4df9af57f.png">

It will help if we vertically align the navigation items on a line:

<img width="416" alt="Screenshot 2020-06-03 at 13 29 24" src="https://user-images.githubusercontent.com/1204802/83631729-4472a000-a59e-11ea-8677-3f6b282952c1.png">

So probably some padding tweaks inside each block item.

The square black inserter — is that a separate PR or could it be part of this one? That will help anchor and balance things too. 

The horizontal "tree-branch" isn't vertically centered with the menu item:

<img width="340" alt="Screenshot 2020-06-03 at 13 31 10" src="https://user-images.githubusercontent.com/1204802/83631894-8ef41c80-a59e-11ea-981a-0c2329e9d945.png">

The vertical tree branch could also be better centered under the icon, but that's minor.

Finally, I do think it's important that we move the arrows closer to each other, to match the mockup, even if that means a little extra focus rectangle work. Might need to use a pseudo element like I did in https://github.com/WordPress/gutenberg/pull/22673. 

Cool work, this one is fast getting there!

adamziel committed:
Remove hover from this PR

adamziel committed:
Apply correct alignment to navigator items

adamziel committed:
Use pseudo elements for focus targets while preserving the large tap target

adamziel committed:
Fix the vertical line alignment

adamziel committed:
Solve conflicts

adamziel commented:
@jasmussen I adjusted the whitespace, removed the hover, and moved the arrows closer together. Ideally I'd like to reuse the same component as the toolbar - let's look into it afterwards.

> The square black inserter — is that a separate PR or could it be part of this one? That will help anchor and balance things too.
It is a separate PR: https://github.com/WordPress/gutenberg/pull/22590

adamziel commented:
Actually after a rebase the alignment is still not quite there, looking into it now.

adamziel commented:
Alright I think it works pretty well now @jasmussen

adamziel committed:
Fix alignment of the descender lines

talldan commented on packages/block-editor/src/components/block-navigation/block.js:
```diff
@@ -96,26 +96,24 @@ export default function BlockNavigationBlock( {
-								clientIds={ [ clientId ] }
-								{ ...props }
-							/>
+							<div className="block-editor-block-navigation-block__mover-cell-alignment-wrapper">
+								<BlockMoverUpButton
+									__experimentalOrientation="vertical"
+									clientIds={ [ clientId ] }
+									{ ...props }
+								/>
+								<BlockMoverDownButton
+									__experimentalOrientation="vertical"
+									clientIds={ [ clientId ] }
+									{ ...props }
+								/>
+							</div>
```
The `props` passed by TreeGridCell to the buttons are used to manage the tab index on focusable elements, so that the whole TreeGrid uses a roving tab index. Moving them both into the same cell so that they use the same props will result in that not working correctly (both buttons will get the same tabIndex).

I mentioned on this [PR](https://github.com/WordPress/gutenberg/pull/22590#discussion_r429753670) that it'd probably be better if the props were passed individually rather than spread as a single `...props` object—making it explicit would more clearly show the connection between the components. I think that's why this issue has happened. The block settings menu seems to need to be fixed too, I'll try to spin up a PR to handle that.

I'm not sure if it's a negative to have two focusable elements within a single cell in this tree grid pattern. If it's ok then it'd probably be easiest to have separate `TreeGridCell` and `TreeGridItem` components, where `TreeGridItem` doesn't have the wrapping `td`.

If not, then the other approach would probably be to use CSS grid. On https://github.com/WordPress/gutenberg/pull/22314/files#diff-3d2c7e20b234263753f1a83b39a15724R31-R33 I experimented with that and it works ok. Though it'd need some work for IE11. I don't think IE11 supports the named grid areas and some of the property names are different.

talldan replied:
Here's a PR that fixes the issue with the Block Settings Menu- https://github.com/WordPress/gutenberg/pull/22885.

adamziel replied:
@talldan good notes! I just added `TreeGridItem` to this PR and it works quite well. The only part that's debatable is how left/right arrows should work when the focus is on these vertical movers. Should the right arrow move the focus from the top mover to the bottom mover? Or should that be the down arrow? Personally I think down arrow would make more sense here.

talldan replied:
@adamziel Yep, agreed. I made an issue for that and it's already had some discussion - https://github.com/WordPress/gutenberg/issues/22316.

It's an issue that exists in the main block toolbar as well. I don't think it's a blocker for this.

The conflict is that left/right makes sense for users who rely only on a screen reader (e.g. users who might have impaired vision), but it's definitely confusing for users using the visual layout of the buttons.

jasmussen commented:
This is really close, it's a pleasure to watch you work!

Few small things:

The block navigator when invoked from the block toolbar is a little big, I recall it being smaller previously?

![big](https://user-images.githubusercontent.com/1204802/83723634-4b052400-a63f-11ea-9466-827c8f03fed3.gif)

Things are looking really good in the navigation block sidebar:

![sidebar](https://user-images.githubusercontent.com/1204802/83723667-55272280-a63f-11ea-801b-a0a8d191a0e1.gif)

In the new "Navigation" screen that's in beta, it looks off for me:

![nav screen](https://user-images.githubusercontent.com/1204802/83723724-66702f00-a63f-11ea-9943-5875a03d90f3.gif)

The padding on the left is too big, and the ellipsis while correctly placed at the end is not vertically centered in the bar. 

Probably all small fixes, but after that, feels good!

adamziel committed:
Fix paddings on the menu management screen

adamziel committed:
Use separate TreeGridItem for mover arrows

adamziel commented:
@jasmussen I fixed the margin issues aside of this one:

> The block navigator when invoked from the block toolbar is a little big, I recall it being smaller previously?

Unfortunately I am not able to reproduce it. Would you mind taking it for a spin if it still shows up on your end?

jasmussen commented:
> Unfortunately I am not able to reproduce it. Would you mind taking it for a spin if it still shows up on your end?

Nope, I can't reproduce anymore. Must've been a cache ghost or something?

<img width="737" alt="Screenshot 2020-06-08 at 11 41 26" src="https://user-images.githubusercontent.com/1204802/84016100-15cd3e80-a97d-11ea-8edd-32db80b8481c.png">

And by the way, NICE. That square plus really helps a lot:

<img width="696" alt="Screenshot 2020-06-08 at 11 41 44" src="https://user-images.githubusercontent.com/1204802/84016127-1d8ce300-a97d-11ea-8897-b30698d7d8e6.png">

adamziel commented:
The inserter is all @noisysocks in https://github.com/WordPress/gutenberg/pull/22590, I just rebased :-) If it looks good, would you mind approving?

talldan reviewed (approved):
Great work here @adamziel. 🎉 

I noticed a few minor style problems when the branch is rebased against `master` that might need to be tweaked :
<img width="270" alt="Screenshot 2020-06-10 at 10 02 37 am" src="https://user-images.githubusercontent.com/677833/84219162-7639b880-ab02-11ea-873b-86a3e597f47a.png">

- The settings menu icon button is misaligned, might just need some padding adjustment (I think there was a separate PR that adjusted it's alignment)
- Getting the appender lined up with the block icons would be great

In the post editor:
<img width="239" alt="Screenshot 2020-06-10 at 10 11 30 am" src="https://user-images.githubusercontent.com/677833/84219337-dcbed680-ab02-11ea-99c7-dc06ccedef2e.png">

- There's a slight overlap between the focus ring and the descending line.

And then I noticed there was a comment about removing some `!important`s from the css. Hopefully those should all be relatively easy things though.

adamziel committed:
Improve alignment

adamziel committed:
Update alignment of descender lines
